subprojects/core/src/main/java/org/gradle/plugin/management/internal/autoapply/AutoAppliedPluginHandler.java

```diff
@@ -16,28 +16,77 @@
 
 package org.gradle.plugin.management.internal.autoapply;
 
+import org.gradle.api.Project;
+import org.gradle.plugin.management.PluginRequest;
+import org.gradle.plugin.management.internal.ArgumentLoadedPluginRequest;
+import org.gradle.plugin.management.internal.PluginRequestInternal;
 import org.gradle.plugin.management.internal.PluginRequests;
 
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
 /**
  * <p>
  * Certain plugins are important enough that Gradle should auto-apply them when it is clear
  * that the user is trying to use it. For instance, when the user uses the <code>--scan</code> option, it
  * is clear they expect the build scan plugin to be applied.
  * </p>
- *
  * Auto-application of a plugin is skipped in the following situations, so the user can adjust the version they want:
- *
  * <ul>
  * <li> The plugin was already applied (e.g. through an init script)
  * <li> The plugin was already requested in the <code>plugins {}</code> block </li>
  * <li> The plugin was already requested in the <code>buildscript {}</code> block </li>
  * </ul>
+ * <p>
+ * Gradle also allows for automatically loading plugins via a system property, this class
+ * understands how to discover these plugin requests as well.
  */
 public interface AutoAppliedPluginHandler {
+    String INIT_PROJECT_SPEC_SUPPLIERS_PROP = "org.gradle.internal.buildinit.projectspecs";
 
     /**
      * Returns plugin requests that should be auto-applied
      * based on user requests, the current build invocation and the given target.
      */
     PluginRequests getAutoAppliedPlugins(PluginRequests initialRequests, Object pluginTarget);
+
+    /**
+     * A static util class responsible for gathering {@link PluginRequest}s added outside of any build script.
+     * <p>
+     * These originate from an id + version pair in the form of {@code id:version}.  These are currently
+     * parsed from a system property {@link #INIT_PROJECT_SPEC_SUPPLIERS_PROP}.
+     */
+    static PluginRequests getArgumentLoadedPlugins() {
+        String propValue = System.getProperty(INIT_PROJECT_SPEC_SUPPLIERS_PROP);
```

Comment:
💭 It feels extremely intrusive to couple general plugin application functionality to `gradle init` extensibility

Reply:
I agree, and we've talked about expanding this into a more general mechanism for loading (and optionally applying) plugins.  The coupling is currently entirely in the name of the property and the name of this static field.  

There isn't a requirement that something loaded in this manner is required to be used by init, or that something used by init is required to be loaded in this manner (though there currently isn't any other way to do it, outside of a custom init script).

Keeping these 2 mechanisms separate was something I was keeping in mind.  If you find anything other than these names that need to be separated, please point it out here.

Reply:
The fact that plugins are only *loaded* and not *applied* also strike me as odd.

Reply:
For purposes of `init`, we don't want to apply 3rd-party (or any) plugins to empty projects though.

Reply:
The fact that this happens in `AutoAppliedPluginHandler` which name implies _applying_ the plugins makes this looks dodgy to me. Couldn't this be handled by some first class object instead of adding the support in `AutoAppliedPluginHandler`?

Reply:
That's true, I suppose it doesn't really fit here.  I'll make an `ArgumentLoadedPluginHandler` and do some adjustments.

Reply:
I extracted this mechanism to a new type and made some small adjustments in [this commit](https://github.com/gradle/gradle/pull/30098/commits/b3ad6778302ceb35a041038f3ba7e27fe55cb6b9).  Lots of renames, and adjustments to use the new `PluginHandler` interface that extends the auto-applied handler and the argument-supplied handler.